jujugui/static/gui/src/app/store/env/bakery.js

```diff
@@ -35,302 +35,313 @@ YUI.add('juju-env-bakery', function(Y) {
-       @param {String} The third party location where to discharge
-       @param {Function} The macaroon to be discharge
-       @param {Function} The request to be sent again in case of
-       successful authentication
-       @param {Function} The callback failure in case of wrong authentication
-       @return {Object} The asynchronous request instance.
-       */
-      _obtainThirdPartyDischarge: function (location,
-                                            thirdPartyLocation, condition,
-                                            successCallback, failureCallback) {
-        thirdPartyLocation += '/discharge';
-        var headers = {
-          'Bakery-Protocol-Version': 1,
-          'Content-Type': 'application/x-www-form-urlencoded'
-        };
-        var content = 'id=' + encodeURI(condition) +
-          '&location=' + encodeURI(location);
-        return this.webhandler.sendPostRequest(
-          thirdPartyLocation,
-          headers, content, null, null, false, null,
-          this._requestHandler.bind(
-            this,
-            this._exportMacaroon.bind(this, successCallback, failureCallback),
-            this._interact.bind(this, successCallback, failureCallback)
-          )
-        );
-      },
+      } catch (ex) {
+        failureCallback(ex.message);
+      }
+    },
+
+    /**
+     Process the discharged macaroon and call the end point to be able to set
+     a cookie for the origin domain only when an auth cookie path is
+     provided, then call the original function.
+
+     @method _processDischarges
+     @param {Function} The request to be sent again in case of
+            successful authentication.
+     @param {Function} The callback failure in case of wrong authentication.
+     @param {[Macaroon]} The macaroons being discharged.
+     @return {undefined} Nothing.
+     */
+    _processDischarges: function (requestFunction, failureCallback,
+                                  discharges) {
+      var jsonMacaroon;
+      try {
+        jsonMacaroon = macaroon.export(discharges);
+      } catch (ex) {
+        failureCallback(ex.message);
+      }
+      var content = JSON.stringify({'Macaroons': jsonMacaroon});
+      if (this.setCookiePath === undefined) {
+        this._successfulDischarges(requestFunction, jsonMacaroon);
+        return;
+      }
+      this.webhandler.sendPutRequest(
+        this.setCookiePath,
+        null, content, null, null, true, null,
+        this._requestHandler.bind(
+          this,
+          this._successfulDischarges.bind(this, requestFunction, jsonMacaroon),
+          failureCallback
+        )
+      );
+    },
+
+    /**
+     Process successful discharge by setting Macaroons Cookie
+     and invoke the original request.
+
+     @method _successfulDischarges
+     @param {Function} The path where to send put request
+            to set the cookie back.
+     @param {Object} an exported Macaroon.
+     @return {undefined} Nothing.
```

Comment:
I think we only need `@return` parts in the docstring when the function actually returns something.